ssl/statem/statem_srvr.c

```diff
@@ -939,8 +939,24 @@ WORK_STATE ossl_statem_server_post_work(SSL *s, WORK_STATE wst)
         break;
 
     case TLS_ST_SW_SESSION_TICKET:
-        if (SSL_IS_TLS13(s) && statem_flush(s) != 1)
+        if (SSL_IS_TLS13(s) && statem_flush(s) != 1) {
+#if defined(EPIPE) && defined(ECONNRESET)
+            if (SSL_get_error(s, 0) == SSL_ERROR_SYSCALL
+                    && (get_last_sys_error() == EPIPE
+                        || get_last_sys_error() == ECONNRESET)) {
```

Comment:
I think you will find that this is too late to be calling get_last_sys_error in that the underlying functions hit state that is reset in the calls in code paths from the read/write through until here. You need to look at get_last_socket_error() in its usage and reference the error values that are saved at the point of the IO call. That has to be checked.

There is also the function BIO_sock_non_fatal_error (poorly named) where it has the set of things for non-blocking IO - and it would make sense to me to follow the same style and have a function we pass the error to which makes the determination so we can add/remove items.

Reply:
> I think you will find that this is too late to be calling get_last_sys_error in that the underlying functions hit state that is reset in the calls in code paths from the read/write through until here.

I don't think so. We don't try and write anything to the underlying socket until we call `statem_flush()`. Up until that point we just buffer anything we attempt to write. The `statem_flush()` causes the buffering BIO to write out all of its data to the underlying socket - and then we immediately test the value of `get_last_sys_error()`. So there really shouldn't be any code paths that reset this state.

> You need to look at get_last_socket_error() in its usage and reference the error values that are saved at the point of the IO call

The difference between `get_last_socket_error()` and `get_last_sys_error` is that, on Windows, the former calls `GetLastError()` whilst the latter calls `WSAGetLastError()`. I'm not enough of a Windows programmer to know if one is preferred over the other. I don't think there are any error values saved immediately at the point of the IO call. But as long as we don't do any more system calls between where the error is and the point where we check it then that shouldn't be necessary.

> There is also the function BIO_sock_non_fatal_error (poorly named) where it has the set of things for non-blocking IO - and it would make sense to me to follow the same style and have a function we pass the error to which makes the determination so we can add/remove items.

Ok - I can do that

Reply:
> The difference between get_last_socket_error() and get_last_sys_error is that, on Windows, the former calls GetLastError() whilst the latter calls WSAGetLastError()

Sorry, its the other way around.